File: web/src/lib/workflow-editor/visibility.ts
import type { PropertyDefinition, Visibility, Condition } from '$lib/api/generated/models';

/**
 * Whether a property is shown, by exactly the rule the server evaluates.
 *
 * This is a port, not a reimplementation: `internal/property/visibility.go` and
 * this file are checked against the same fixture, because two implementations
 * of one rule drift — that is not a risk, it is a certainty — and the fixture is
 * the only thing that keeps the panel and the compiler agreeing about whether a
 * workflow can be saved.
 *
 * A hidden parent does not suppress its children: every property is evaluated
 * independently against the node's stored parameters. n8n behaves the same way,
 * and a cascade would hide parameters an imported workflow legitimately sets.
 */
export function visible(
	rule: Visibility | undefined,
	parameters: Record<string, unknown>,
	typeVersion = ''
): boolean {
	const show = rule?.show ?? [];
	const hide = rule?.hide ?? [];
	if (show.length === 0 && hide.length === 0) return true;

	for (const condition of show) {
		const [value, present] = controlling(condition.key, parameters, typeVersion);
		// The escape hatch, and it is in the show branch only. If the
		// controlling parameter holds an expression, the dependent property is
		// always shown: nothing can know at edit time what it will evaluate to,
		// and hiding a field the user may need is worse than showing one they
		// may not.
		if (present && isExpressionMarker(value)) continue;
		if (!matches(condition, value, present)) return false;
	}

	for (const condition of hide) {
		const [value, present] = controlling(condition.key, parameters, typeVersion);
		// An absent value never hides.
		if (!present || value === null || value === undefined) continue;
		if (matches(condition, value, present)) return false;
	}
	return true;
}

/**
 * Fills in the parameters a node never stored.
 *
 * A property the user never touched has its declared default, and every
 * visibility rule has to be evaluated against that. Without it, a rule reading
 * `mode` on a node whose `mode` was never written sees nothing and hides a
 * field the user is looking at.
 */
export function withDefaults(
	properties: PropertyDefinition[],
	parameters: Record<string, unknown>
): Record<string, unknown> {
	const filled: Record<string, unknown> = { ...parameters };
	for (const property of properties) {
		if (property.default === undefined || property.default === null) continue;
		if (!(property.key in filled)) filled[property.key] = property.default;
	}
	return filled;
}

/** Whether one property is shown, honouring the visibleWhen shorthand. */
export function propertyVisible(
	property: PropertyDefinition,
	parameters: Record<string, unknown>,
	typeVersion = ''
): boolean {
	const rule = property.displayOptions;
	if (rule && ((rule.show?.length ?? 0) > 0 || (rule.hide?.length ?? 0) > 0)) {
		return visible(rule, parameters, typeVersion);
	}
	const shorthand = property.visibleWhen ?? [];
	if (shorthand.length === 0) return true;
	return visible(
		{ show: shorthand.map((condition) => ({ key: condition.key, values: [condition.equals] })) },
		parameters,
		typeVersion
	);
}

function controlling(
	key: string,
	parameters: Record<string, unknown>,
	typeVersion: string
): [unknown, boolean] {
	// The pseudo-key gating on the node's type version.
	if (key === '@version') return [typeVersion, typeVersion !== ''];
	return [parameters[key], Object.prototype.hasOwnProperty.call(parameters, key)];
}

function isExpressionMarker(value: unknown): boolean {
	if (typeof value !== 'object' || value === null || Array.isArray(value)) return false;
	const record = value as Record<string, unknown>;
	return record.mode === 'expression' && typeof record.value === 'string';
}

function matches(condition: Condition, value: unknown, present: boolean): boolean {
	const operator = condition.operator || 'eq';
	const operands = condition.values ?? [];

	if (operator === 'exists') return present && value !== null && value !== undefined;
	if (operator === 'between') {
		if (operands.length !== 2) return false;
		return compare(value, operands[0]) >= 0 && compare(value, operands[1]) <= 0;
	}
	if (operator === 'not') {
		// Negation is over the whole list: "not one of these".
		return !operands.some((operand) => equal(value, operand));
	}
	return operands.some((operand) => apply(operator, value, operand));
}

function apply(operator: string, value: unknown, operand: unknown): boolean {
	switch (operator) {
		case 'eq':
			return equal(value, operand);
		case 'gt':
			return compare(value, operand) > 0;
		case 'gte':
			return compare(value, operand) >= 0;
		case 'lt':
			return compare(value, operand) < 0;
		case 'lte':
			return compare(value, operand) <= 0;
		case 'startsWith':
			return text(value).startsWith(text(operand));
		case 'endsWith':
			return text(value).endsWith(text(operand));
		case 'includes':
			if (Array.isArray(value)) return value.some((entry) => equal(entry, operand));
			return text(value).includes(text(operand));
		case 'regex':
			try {
				return new RegExp(text(operand)).test(text(value));
			} catch {
				return false;
			}
		default:
			return false;
	}
}

/**
 * Compares by value, not by identity.
 *
 * This is what the previous `===` got wrong: it never matched an object or an
 * array, so a condition on anything but a primitive was silently always false.
 * A JSON number and its string form compare equal, because they are the same
 * value to a user reading a dropdown.
 */
function equal(value: unknown, operand: unknown): boolean {
	const left = numeric(value);
	const right = numeric(operand);
	if (left !== null && right !== null) return left === right;
	if (left !== null || right !== null) return text(value) === text(operand);
	if (typeof value === 'object' && typeof operand === 'object') {
		return JSON.stringify(sorted(value)) === JSON.stringify(sorted(operand));
	}
	return value === operand;
}

function compare(value: unknown, operand: unknown): number {
	const left = numeric(value);
	const right = numeric(operand);
	if (left === null || right === null) {
		const a = text(value);
		const b = text(operand);
		return a < b ? -1 : a > b ? 1 : 0;
	}
	return left < right ? -1 : left > right ? 1 : 0;
}

function numeric(value: unknown): number | null {
	if (typeof value === 'number' && Number.isFinite(value)) return value;
	if (typeof value === 'string' && value.trim() !== '') {
		const parsed = Number(value);
		return Number.isFinite(parsed) ? parsed : null;
	}
	return null;
}

function text(value: unknown): string {
	if (value === null || value === undefined) return '';
	if (typeof value === 'string') return value;
	return String(value);
}

/** Key-sorted so two equal objects stringify identically. */
function sorted(value: unknown): unknown {
	if (Array.isArray(value)) return value.map(sorted);
	if (typeof value !== 'object' || value === null) return value;
	const record = value as Record<string, unknown>;
	const result: Record<string, unknown> = {};
	for (const key of Object.keys(record).sort()) result[key] = sorted(record[key]);
	return result;
}

File: internal/property/visibility.go
package property

import (
	"fmt"
	"reflect"
	"regexp"
	"strconv"
	"strings"
)

// VersionKey is the pseudo-key a condition uses to gate on the node's type
// version, so one definition can present different parameters per version.
//
// n8n also has `@feature` and `@tool`. They gate on instance features that have
// no KilasFlow equivalent, so accepting them would mean inventing semantics —
// they are refused by name at registration instead.
const VersionKey = "@version"

// Operator compares a controlling value against a condition's operands.
type Operator string

const (
	OperatorEquals     Operator = "eq"
	OperatorNotEquals  Operator = "not"
	OperatorGreater    Operator = "gt"
	OperatorGreaterEq  Operator = "gte"
	OperatorLess       Operator = "lt"
	OperatorLessEq     Operator = "lte"
	OperatorBetween    Operator = "between"
	OperatorStartsWith Operator = "startsWith"
	OperatorEndsWith   Operator = "endsWith"
	OperatorIncludes   Operator = "includes"
	OperatorRegex      Operator = "regex"
	OperatorExists     Operator = "exists"
)

// KnownOperators is the closed set.
func KnownOperators() []Operator {
	return []Operator{
		OperatorEquals, OperatorNotEquals,
		OperatorGreater, OperatorGreaterEq, OperatorLess, OperatorLessEq,
		OperatorBetween, OperatorStartsWith, OperatorEndsWith,
		OperatorIncludes, OperatorRegex, OperatorExists,
	}
}

// Condition is one key and the values it accepts.
//
// Within a single condition the values are OR'd: "operation is one of send,
// sendPhoto or sendDocument" is one condition with three values. Across
// conditions in a `show` group they are AND'd. That asymmetry is n8n's and is
// what makes the common shapes expressible without nesting.
type Condition struct {
	Key string `json:"key"`
	// Values are the literals this key may hold. Empty with an operator means
	// the operator needs no operand, as `exists` does.
	Values []any `json:"values,omitempty"`
	// Operator defaults to equality.
	Operator Operator `json:"operator,omitempty"`
}

// Visibility decides whether a property is shown.
//
// Under Show, **every** listed key must match. Under Hide, **any** matching key
// hides — but only when the controlling parameter actually has a value, so an
// absent value never hides. Both rules are n8n's, copied deliberately: a node
// imported from n8n presents its parameters by them, and a near-miss here shows
// the wrong fields on every imported node.
type Visibility struct {
	Show []Condition `json:"show,omitempty"`
	Hide []Condition `json:"hide,omitempty"`
}

// IsEmpty reports a property with no visibility rule, which is always shown.
func (visibility Visibility) IsEmpty() bool {
	return len(visibility.Show) == 0 && len(visibility.Hide) == 0
}

// ExpressionMarker reports a value that is a KilasFlow expression.
//
// n8n marks one with a leading `=` on a plain string; KilasFlow uses the
// explicit object. The marker differs, the rule does not.
func ExpressionMarker(value any) bool {
	object, isObject := value.(map[string]any)
	if !isObject {
		return false
	}
	mode, _ := object["mode"].(string)
	if mode != "expression" {
		return false
	}
	_, hasValue := object["value"].(string)
	return hasValue
}

// Visible evaluates a property's rule against a node's stored parameters.
//
// A hidden parent does **not** suppress its children. Every property is
// evaluated independently against the stored parameters, so a property whose
// controlling parameter is itself hidden is still evaluated on that parameter's
// stored value. Imported nodes depend on this — n8n behaves the same way, and a
// cascade would hide parameters an imported workflow legitimately sets.
func Visible(visibility Visibility, parameters map[string]any, typeVersion string) bool {
	if visibility.IsEmpty() {
		return true
	}

	for _, condition := range visibility.Show {
		value, present := controllingValue(condition.Key, parameters, typeVersion)
		// The escape hatch, and it lives in the show branch only. If the
		// controlling parameter holds an expression, the dependent property is
		// always shown: nothing can know at edit time what that expression will
		// evaluate to, and hiding a field the user may need is worse than
		// showing one they may not.
		if present && ExpressionMarker(value) {
			continue
		}
		if !matches(condition, value, present) {
			return false
		}
	}

	for _, condition := range visibility.Hide {
		value, present := controllingValue(condition.Key, parameters, typeVersion)
		// An absent value never hides. n8n checks the value list is non-empty
		// before comparing, which is the same rule stated from the other side:
		// a property is not hidden by a parameter nobody has set.
		if !present || value == nil {
			continue
		}
		if matches(condition, value, present) {
			return false
		}
	}
	return true
}

func controllingValue(key string, parameters map[string]any, typeVersion string) (any, bool) {
	if key == VersionKey {
		return typeVersion, typeVersion != ""
	}
	value, present := parameters[key]
	return value, present
}

// matches applies one condition. Values within it are OR'd.
func matches(condition Condition, value any, present bool) bool {
	operator := condition.Operator
	if operator == "" {
		operator = OperatorEquals
	}
	if operator == OperatorExists {
		return present && value != nil
	}
	if operator == OperatorBetween {
		if len(condition.Values) != 2 {
			return false
		}
		return compareNumbers(value, condition.Values[0]) >= 0 &&
			compareNumbers(value, condition.Values[1]) <= 0
	}
	if operator == OperatorNotEquals {
		// Negation is over the whole list: "not one of these", not "not the
		// first and not the second considered separately".
		for _, operand := range condition.Values {
			if equal(value, operand) {
				return false
			}
		}
		return true
	}
	for _, operand := range condition.Values {
		if applyOperator(operator, value, operand) {
			return true
		}
	}
	return false
}

func applyOperator(operator Operator, value, operand any) bool {
	switch operator {
	case OperatorEquals:
		return equal(value, operand)
	case OperatorGreater:
		return compareNumbers(value, operand) > 0
	case OperatorGreaterEq:
		return compareNumbers(value, operand) >= 0
	case OperatorLess:
		return compareNumbers(value, operand) < 0
	case OperatorLessEq:
		return compareNumbers(value, operand) <= 0
	case OperatorStartsWith:
		return strings.HasPrefix(text(value), text(operand))
	case OperatorEndsWith:
		return strings.HasSuffix(text(value), text(operand))
	case OperatorIncludes:
		if list, isList := value.([]any); isList {
			for _, entry := range list {
				if equal(entry, operand) {
					return true
				}
			}
			return false
		}
		return strings.Contains(text(value), text(operand))
	case OperatorRegex:
		pattern, err := regexp.Compile(text(operand))
		if err != nil {
			return false
		}
		return pattern.MatchString(text(value))
	default:
		return false
	}
}

// equal compares by value, not by identity.
//
// This is precisely what the previous `===` got wrong: it never matched an
// object or an array, so a condition on anything but a primitive was silently
// always false. Numbers are compared as float64 because a value arriving
// through encoding/json is one while a condition written in Go is an int, and
// they must agree.
func equal(value, operand any) bool {
	left, leftNumeric := numeric(value)
	right, rightNumeric := numeric(operand)
	if leftNumeric && rightNumeric {
		return left == right
	}
	// A JSON number and its string form are the same value to a user reading a
	// dropdown, so they compare equal — otherwise a select whose options are
	// strings can never gate on a numeric parameter.
	if leftNumeric != rightNumeric {
		return text(value) == text(operand)
	}
	return reflect.DeepEqual(value, operand)
}

func compareNumbers(value, operand any) int {
	left, leftNumeric := numeric(value)
	right, rightNumeric := numeric(operand)
	if !leftNumeric || !rightNumeric {
		return strings.Compare(text(value), text(operand))
	}
	switch {
	case left < right:
		return -1
	case left > right:
		return 1
	default:
		return 0
	}
}

func numeric(value any) (float64, bool) {
	switch typed := value.(type) {
	case float64:
		return typed, true
	case float32:
		return float64(typed), true
	case int:
		return float64(typed), true
	case int64:
		return float64(typed), true
	case string:
		parsed, err := strconv.ParseFloat(strings.TrimSpace(typed), 64)
		return parsed, err == nil
	default:
		return 0, false
	}
}

func text(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	case nil:
		return ""
	case float64:
		return strconv.FormatFloat(typed, 'f', -1, 64)
	case bool:
		return strconv.FormatBool(typed)
	default:
		return fmt.Sprint(typed)
	}
}

// ValidateVisibility refuses a rule this server cannot honour.
func ValidateVisibility(visibility Visibility) error {
	for _, group := range [][]Condition{visibility.Show, visibility.Hide} {
		for _, condition := range group {
			if strings.TrimSpace(condition.Key) == "" {
				return fmt.Errorf("a visibility condition names no key")
			}
			if strings.HasPrefix(condition.Key, "@") && condition.Key != VersionKey {
				// Accepting and ignoring would present the wrong fields
				// silently; naming it is the honest failure.
				return fmt.Errorf("visibility key %q is not supported; only %s is", condition.Key, VersionKey)
			}
			if condition.Operator == "" {
				continue
			}
			var known bool
			for _, operator := range KnownOperators() {
				if condition.Operator == operator {
					known = true
				}
			}
			if !known {
				return fmt.Errorf("visibility operator %q is not supported", condition.Operator)
			}
		}
	}
	return nil
}
